opendevin/controller/agent_controller.py

```diff
@@ -81,101 +75,74 @@ def __init__(
             EventStreamSubscriber.AGENT_CONTROLLER, self.on_event
         )
         self.max_iterations = max_iterations
-
-        self.runtime = ServerRuntime(sandbox=sandbox, sid=self.id)
         self.max_chars = max_chars
-
-        # Initialize agent-required plugins for sandbox (if any)
-        self.runtime.init_sandbox_plugins(agent.sandbox_plugins)
-
-        if isinstance(agent, CodeActAgent) and not isinstance(
-            self.runtime.sandbox, DockerSSHBox
-        ):
-            logger.warning(
-                'CodeActAgent requires DockerSSHBox as sandbox! Using other sandbox that are not stateful (LocalBox, DockerExecBox) will not work properly.'
-            )
+        self.agent_task = asyncio.create_task(self._start_step_loop())
 
     async def close(self):
         if self.agent_task is not None:
             self.agent_task.cancel()
         self.event_stream.unsubscribe(EventStreamSubscriber.AGENT_CONTROLLER)
-        self.runtime.sandbox.close()
-        self.runtime.browser.close()
         await self.set_agent_state_to(AgentState.STOPPED)
 
-    def update_state_for_step(self, i):
-        self.state.iteration = i
-        self.state.background_commands_obs = self.runtime.get_background_obs()
+    def update_state_before_step(self):
+        self.state.iteration += 1
 
     def update_state_after_step(self):
         self.state.updated_info = []
 
-    async def add_error_to_history(self, message: str):
-        await self.add_history(NullAction(), ErrorObservation(message))
+    async def report_error(self, message: str):
+        await self.event_stream.add_event(ErrorObservation(message), EventSource.AGENT)
 
-    async def add_history(
-        self, action: Action, observation: Observation, add_to_stream=True
-    ):
-        if not isinstance(action, Action):
-            raise TypeError(
-                f'action must be an instance of Action, got {type(action).__name__} instead'
-            )
-        if not isinstance(observation, Observation):
-            raise TypeError(
-                f'observation must be an instance of Observation, got {type(observation).__name__} instead'
-            )
+    async def add_history(self, action: Action, observation: Observation):
+        if isinstance(action, NullAction) and isinstance(observation, NullObservation):
+            return
```

Comment:
This makes sense and will be good to have, but it breaks something with _is_stuck() checks, if... if it's still working at the moment, which I'm not sure. It relies on (Action, Observation) tuples, with a special case for NullObservation.

FWIW I did see recently _is_stuck() working with CodeAct, so it was working at some point with all. Including thoughts with NullObservations.

IMHO we can let it be outside this PR, I'll look at it again after this refactoring is done; alternatively we can replace the check for NullObservation with any identical Observation, and maybe increase it to 5 or 6 instead of 3.
(...and maybe avoid MessageAction with source user, if it's there... I don't know, it may need some tweaks.)

Reply:
Good thoughts--I'm looking at the loop logic now and I agree, it could probably use some tweaking. 

Here's an example of a recent loop we didn't catch: https://opendevin.slack.com/archives/C06U8UTKSAD/p1715691234611019

It's funky because it's _alternating_ between two different actions, instead of repeating the same action...

Reply:
Wow, I thought about that before, but I don't think I saw it myself, they were still varying _a little_, unlike that one.